<!doctype htboml>
<!-- paulirish.com/2008/conditional-stylesheets-vs-css-hacks-answer-neither/ -->
<!--[if lt IE 7]> <html class="no-js ie6 oldie" lang="en"> <![endif]-->
<!--[if IE 7]>    <html class="no-js ie7 oldie" lang="en"> <![endif]-->
<!--[if IE 8]>    <html class="no-js ie8 oldie" lang="en"> <![endif]-->
<!-- Consider adding an manifest.appcache: h5bp.com/d/Offline -->
<!--[if gt IE 8]><!-->
<html class="no-js" lang="en">
<!--<![endif]-->
<head>
<meta charset="utf-8">

<!-- Use the .htaccess and remove these lines to avoid edge case issues.
 More info: h5bp.com/b/378 -->
<!-- <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1"> -->
<!-- Not yet implemented -->

<title>Gene Search &lt; Ensembl &lt; EMBL-EBI</title>
<meta name="description" content="EMBL-EBI">
<!-- Describe what this page is about -->
<meta name="keywords" content="bioinformatics, europe, institute">
<!-- A few keywords that relate to the content of THIS PAGE (not the whol project) -->
<meta name="author" content="EMBL-EBI">
<!-- Your [project-name] here -->

<!-- Mobile viewport optimized: j.mp/bplateviewport -->
<meta name="viewport" content="width=device-width,initial-scale=1">

<!-- Place favicon.ico and apple-touch-icon.png in the root directory: mathiasbynens.be/notes/touch-icons -->

<!-- CSS: implied media=all -->
<!-- CSS concatenated and minified via ant build script-->
<link rel="stylesheet" href="css/bootstrap.min.css" type="text/css"
	media="screen">
<link rel="stylesheet"
	href="http://www.ebi.ac.uk/web_guidelines/css/compliance/develop/boilerplate-style.css">
<link rel="stylesheet"
	href="http://www.ebi.ac.uk/web_guidelines/css/compliance/develop/ebi-global.css"
	type="text/css" media="screen">
<link rel="stylesheet"
	href="http://www.ebi.ac.uk/web_guidelines/css/compliance/develop/ebi-visual.css"
	type="text/css" media="screen">
<link rel="stylesheet"
	href="http://www.ebi.ac.uk/web_guidelines/css/compliance/develop/984-24-col-fluid.css"
	type="text/css" media="screen">
<link rel="stylesheet" type="text/css"
	href="//cdn.datatables.net/1.10.11/css/jquery.dataTables.css">
<link rel="stylesheet"
	href="https://ajax.googleapis.com/ajax/libs/jqueryui/1.11.4/themes/smoothness/jquery-ui.css">
<link rel="stylesheet" type="text/css"
	href="https://cdnjs.cloudflare.com/ajax/libs/select2/4.0.2/css/select2.min.css">
<link rel="stylesheet" type="text/css" href="css/select2-bootstrap.css">
<!-- you can replace this with [projectname]-colours.css. See http:http://frontier.ebi.ac.uk/web/style/colour for details of how to do this -->
<!-- also inform ES so we can host your colour palette file -->
<link rel="stylesheet"
	href="http://www.ebi.ac.uk/web_guidelines/css/compliance/develop/embl-petrol-colours.css"
	type="text/css" media="screen">

<!-- for production the above can be replaced with -->
<!--
    <link rel="stylesheet" href="http://www.ebi.ac.uk/web_guidelines/css/compliance/mini/ebi-fluid-embl.css">
    -->


<!-- end CSS-->


<!-- All JavaScript at the bottom, except for Modernizr / Respond.
Modernizr enables HTML5 elements & feature detects; Respond is a polyfill for min/max-width CSS3 Media Queries
For optimal performance, use a custom Modernizr build: www.modernizr.com/download/ -->

<!-- Full build -->
<!-- <script src="http://www.ebi.ac.uk/web_guidelines/js/libs/modernizr.minified.2.1.6.js"></script> -->

<!-- custom build (lacks most of the "advanced" HTML5 support -->
<script
	src="http://www.ebi.ac.uk/web_guidelines/js/libs/modernizr.custom.49274.js"></script>


</head>

<body class="level2">
	<!-- add any of your classes or IDs -->
	<div id="skip-to">
		<ul>
			<li><a href="#content">Skip to main content</a></li>
			<li><a href="#local-nav">Skip to local navigation</a></li>
			<li><a href="#global-nav">Skip to EBI global navigation menu</a></li>
			<li><a href="#global-nav-expanded">Skip to expanded EBI
					global navigation menu (includes all sub-sections)</a></li>
		</ul>
	</div>

	<div id="wrapper" class="container_24">
		<header>
			<div id="global-masthead" class="masthead grid_24">
				<!--This has to be one line and no newline characters-->
				<a href="http://www.ebi.ac.uk/" title="Go to the EMBL-EBI homepage"><img
					src="http://www.ebi.ac.uk/web_guidelines/images/logos/EMBL-EBI/EMBL_EBI_Logo_white.png"
					alt="EMBL European Bioinformatics Institute"></a>

				<nav>
					<ul id="global-nav">
						<!-- set active class as appropriate -->
						<li id="services" class=" first active"><a
							href="http://www.ebi.ac.uk/services" title="Services">Services</a></li>
						<li id="research" class=""><a
							href="http://www.ebi.ac.uk/research" title="Research">Research</a></li>
						<li id="training" class=""><a
							href="http://www.ebi.ac.uk/training" title="Training">Training</a></li>
						<li id="industry" class=""><a
							href="http://www.ebi.ac.uk/industry" title="Industry">Industry</a></li>
						<li id="about" class=" last"><a
							href="http://www.ebi.ac.uk/about" title="About us">About us</a></li>
					</ul>
				</nav>

			</div>
			<div id="local-masthead" class="masthead grid_24 nomenu">

				<!-- local-title -->
				<!-- NB: for additional title style patterns, see http://frontier.ebi.ac.uk/web/style/patterns -->

				<div class=" logo-title-strapline" id="local-title">
					<a href="/" title="Back to Ensembl Gene Search homepage"><img
						src="images/enslogo.png" alt="Ensembl logo" width="64" height="64"></a>
					<span>
						<h1>
							<a href="/" title="Back to Ensembl Gene Search homepage">Ensembl
								Gene Search</a>
						</h1> <br>
						<p>Finding genes in Ensembl, fast</p>
					</span>
				</div>

				<!-- /local-title -->


				<!-- local-nav -->
				<nav>
					<ul class="grid_24 nav nav-pills" id="local-nav">
						<li class="first active"><a data-toggle="pill" href="#usage">API
								Usage</a></li>
						<li><a data-toggle="pill" href="#examples">API Examples</a></li>
						<li class="last"><a data-toggle="pill" href="#aboutegs">About
								Ensembl Gene Search</a></li>
						<li class="functional"><a href="#" class="icon icon-static"
							data-icon="\">Feedback</a></li>
					</ul>
				</nav>
				<!-- /local-nav -->

			</div>
		</header>

		<div id="content" role="main" class="grid clearfix">

			<div class="tab-content">
				<div class="tab-pane fade in active" class="show" id="usage">
					<h3>API Usage</h3>
					<p>
						The service is available on
						<code>api</code>
						with service descriptors from <a href='api/application.wadl'><code>api/application.wadl</code></a>
					</p>

					<p>
						The application supports GZIP compression of responses where the
						client header has
						<code>Accept-Encoding:gzip</code>
						set.
					</p>

					<p>
						Note that the majority of endpoints described below are
						implemented using the same pattern for the following objects
						(denoted by
						<code>{object}</code>
						in the URL):
					<ul>
						<li>genes</li>
						<li>transcripts</li>
						<li>sequences</li>
						<li>genomes</li>
						<li>variants</li>
						<li>expression</li>
					</ul>
					</p>

					<h4>/api/health
					</h4>

					<h4>/api/{object}/health
					</h4>

					<p>
						These can be used to check if the server and its component
						searches are running properly. The response is a hash containing
						the key
						<code>status</code>
						which can be
						<code>UP</code>
						or
						<code>DOWN</code>
						. The top level health service hash also contains the status of
						each component search e.g.
						<code> { "status": "DOWN", "GENOMES": "up", "GENES": "up",
							"TRANSCRIPTS": "up", "HOMOLOGUES": "up", "VARIANTS": "up",
							"SEQUENCES": "up", "EXPRESSION": "down",
							"EXPRESSION_EXPERIMENTS": "down", "EXPRESSION_ANALYTICS": "down"
							} </code>
					</p>

					<h5>GET /api/{object}/info/
					</h5>

					<p>Returns complete details for the specified datatype e.g.
						<a href='api/genes/info'><code>api/genes/info</code></a> <a href='api/genomes/info'><code>api/genomes/info</code></a></p>

					<h5>GET /api/{object}/info/fields
					</h5>

					<p>Returns all fields for the specified datatype e.g.
						<a href='api/genes/info/fields'><code>api/genes/info/fields</code></a></p>


					<h4>/api/genes/query
						and /api/genomes/query
					</h4>

					<p>
						The
						<code>/api/genes/query</code>
						and
						<code>/api/genomes/query</code>
						endpoints allow the first n hits and a count plus facets to be
						returned as JSON. Both endpoints operate in the same way.
					</p>

					<h5>GET</h5>

					<p>
						The
						<code>/api/genes/query</code>
						endpoint accepts the following arguments: -
						<code>query</code>
						- a JSON-formatted string e.g.
						<code>"{"genome":"homo_sapiens","biotype":"protein_coding"}"</code>
						-
						<code>fields</code>
						- a comma-separated list of fields e.g.
						<code>genome,name,description</code>
						.
						<code>*</code>
						can be used as a wildcard e.g.
						<code>*</code>
						to retrieve all fields -
						<code>sort</code>
						- a comma-separated list of fields to sort by, which can be
						preceded by
						<code>+</code>
						or
						<code>-</code>
						to indicate ascending or descending e.g.
						<code>+name,-start</code>
						-
						<code>limit</code>
						- number of hits to return e.g. 100 -
						<code>facets</code>
						- a comma-separated list of fields to facet over e.g.
						<code>facets=genome,biotype</code>
						-
						<code>target</code>
						- optional name of target if genes are not required (e.g.
						<code>transcripts</code>
						,
						<code>translations</code>
						,
						<code>homologues</code>
						,
						<code>sequences</code>
						) -
						<code>targetQuery</code>
						- optional additional queries passed to the secondary target
					</p>

					<p>
						<a
							href='api/genes/query?query={"genome":"nanoarchaeum_equitans_kin4_m"}&amp;limit=5&fields=name,genome&amp;sort=+name,-start&facets=biotype'><code>api/genes/query?query={"genome":"nanoarchaeum_equitans_kin4_m"}&amp;limit=5&amp;fields=name,genome&amp;sort=+name,-start&amp;facets=biotype</code></a>
					</p>

					<p>
						This endpoint returns a JSON-formatted response containing the
						number of hits, the first
						<code>limit</code>
						hits and a list of facets e.g.
						<code> { "resultCount": 598, "results": [ { "genome":
							"nanoarchaeum_equitans_kin4_m", "name": "5S_rRNA", "id":
							"EBG00001184723" }, { "genome": "nanoarchaeum_equitans_kin4_m",
							"name": "5_8S_rRNA", "id": "EBG00001184731" } ], "facets": {
							"biotype": { "tRNA": 38, "protein_coding": 536, "rRNA": 23,
							"snRNA": 1 } } } </code>
					</p>
					
					<h5>POST</h5>

					<p>
						This is as above, but the parameters are supplied as pure JSON:
						<code> curl -H "Content-Type: application/json" -X POST
							--data
							'{"query":{"genome":"nanoarchaeum_equitans_kin4_m"},"limit":10,
							"fields":["name","genome"],"sort":["+name","-start"]}'
							api/genes/query </code>
					</p>

					<h4>/api/{object}/fetch
					</h4>

					<p>
						The
						<code>/api/genes/fetch</code>
						and
						<code>/api/genomes/fetch</code>
						endpoints are intended for bulk retrieval and accepts the
						<code>query</code>
						,
						<code>fields</code>
						,
						<code>target</code>
						and
						<code>targetQuery</code>
						parameters from
						<code>/query</code>
						. It can be invoked via GET and POST and returns the results as a
						JSON object, with the fields
						<code>fields</code>
						(array of objects representing retrieved fields) and
						<code>results</code>
						(array of all matched documents), with no summary or facets
						available.
					</p>

					<p>
						The output format is JSON by default, but can be changed by
						setting either the
						<code>Accept</code>
						request header or
						<code>accept</code>
						parameter to:<ul>
						<li>JSON -
						<code>application/x-javascript</code></li>
						<li>XML -
						<code>application/xml</code></li>
						<li>CSV -
						<code>text/csv</code>
						or
						<code>application/vnd.ms-excel</code></li></ul>
					</p>

					<h5>GET</h5>

					<p>
						<a
							href='api/genes/fetch?query={"genome":"nanoarchaeum_equitans_kin4_m"}&fields=name,genome&amp;sort=+name,-start'><code>api/genes/fetch?query={"genome":"nanoarchaeum_equitans_kin4_m"}&amp;fields=name,genome&amp;sort=+name,-start</code></a>
					</p>

					<h5>POST</h5>

					<p>
						<code> curl -H "Content-Type: application/json" -X POST
							--data
							'{"query":{"genome":"nanoarchaeum_equitans_kin4_m"},"fields":["name","genome"],"sort":["+name","-start"]}'
							api/genes/fetch </code>
					</p>

					<h4>/api/{object}
					</h4>

					<p>
						The
						<code>/api/genes</code>
						and
						<code>/api/genomes</code>
						endpoint are simple endpoints for retrieving complete documents
						for genes as JSON.
					</p>

					<h5>GET (single gene or genome)</h5>

					<p>
						The gene stable ID or genome production name is supplied as a URL
						parameter:
						<code>api/genes/NEQ392</code>
						<br />
						<code>api/genomes/homo_sapiens</code>
						<br />
					</p>

					<p>
						In addition,
						<code>/api/genomes/select</code>
						is provided to allow genomes to be found by name (this is for use
						in a lookup). For instance, to find genomes with any names
						matching
						<code>thermo</code>
						(note that this uses aliases as well as assigned names):
						<code> api/genomes/select?query=thermo </code>
					</p>

					<h5>POST (multiple genes or genomes)</h5>

					<p>
						A JSON document containing a list of gene stable IDs or genome
						production names can be passed via POST:
						<code> curl -H "Content-Type: application/json" -X POST
							--data '["NEQ392","NEQ393"]' api/genes </code>
					</p>

				</div>

				<div class="tab-pane fade in active" class="show" id="examples">
					<h3>API Examples</h3>

					<h4>
						<code>/query</code>
					</h4>

					<h5>Example queries</h5>

					<p>
						Find all genes called BRCA2 in any genome: <br />
						<code>
							<a href='api/genes/query?query={"name":"BRCA2"}'>api/genes/query?query={"name":"BRCA2"}</a>
						</code>
					</p>

					<p>
						To restrict to a given genome: <br />
						<code>
							<a
								href='api/genes/query?query={"name":"BRCA2","genome":"homo_sapiens"}'>api/genes/query?query={"name":"BRCA2","genome":"homo_sapiens"}</a>
						</code>
					</p>

					<p>
						To restrict to a results except in the given genome: <br />
						<code>
							<a
								href='api/genes/query?query={"name":"BRCA2","!genome":"homo_sapiens"}'>api/genes/query?query={"name":"BRCA2","!genome":"homo_sapiens"}</a>
						</code>
					</p>

					<p>
						To restrict to a given taxonomic lineage (40674 is Mammalia): <br />
						<code>
							<a
								href='api/genes/query?query={"name":"BRCA2","lineage":"40674"}'>api/genes/query?query={"name":"BRCA2","lineage":"40674"}</a>
						</code>
					</p>

					<p>
						To find genes annotated with GO terms that are the children of a
						given term use the GO_expanded field: <br />
						<code>
							<a href='api/genes/query?query={"GO_expanded":"GO:0016787"}'>api/genes/query?query={"GO_expanded":"GO:0016787"}</a>
						</code>
					</p>

					<p>
						To find specific GO term annotation (double-stranded break repair
						with mutant phenotype evidence): <br />
						<code>
							<a
								href='api/genes/query?query={"GO":{"evidence":"IMP","term":"GO:0006302"}}'>api/genes/query?query={"GO":{"evidence":"IMP","term":"GO:0006302"}}</a>
						</code>
					</p>

					<p>
						To find genes with a particular PHI-base annotation (reduced
						virulence in wheat) <br />
						<code>
							<a
								href='api/genes/query?query={"PHI":{"host":"4565","phenotype":"reduced virulence"}}'>api/genes/query?query={"PHI":{"host":"4565","phenotype":"reduced
								virulence"}}</a>
						</code>
					</p>

					<p>
						To find genes with a particular UniProt cross-reference: <br />
						<code>
							<a href='api/genes/query?query={"Uniprot/SWISSPROT":["P03886"]}'>api/genes/query?query={"Uniprot/SWISSPROT":["P03886"]}</a>
						</code>
					</p>

					<p>
						To find genes matching any of a list of UniProt records: <br />
						<code>
							<a
								href='api/genes/query?query={"Uniprot/SWISSPROT":["P03886","P03891","P00395","P00403","P03928"]}'>api/genes/query?query={"Uniprot/SWISSPROT":["P03886","P03891","P00395","P00403","P03928"]}</a>
						</code>
					</p>

					<p>
						To find genes matching with different values of a number such as a
						start position: <br />
						<code>
							<a
								href='api/genes/query?query={"genome":"homo_sapiens", "start":"45000-46000"}'>api/genes/query?query={"genome":"homo_sapiens",
								"start":"45000-46000"}</a> <a
								href='api/genes/query?query={"genome":"homo_sapiens", "start":"&gt;45000"}'>api/genes/query?query={"genome":"homo_sapiens",
								"start":"&gt;45000"}</a> <a
								href='api/genes/query?query={"genome":"homo_sapiens", "start":"&lt;=45000"}'>api/genes/query?query={"genome":"homo_sapiens",
								"start":"&lt;=45000"}</a>
						</code>
					</p>

					<p>
						To find genes in a given location or locations: <br />
						<code>
							<a
								href='api/genes/query?query={"genome":"homo_sapiens", "location":"1:45000-96000"}'>api/genes/query?query={"genome":"homo_sapiens",
								"location":"1:45000-96000"}</a> <a
								href='api/genes/query?query={"genome":"homo_sapiens", "location":"1:45000-96000:-1"}'>api/genes/query?query={"genome":"homo_sapiens",
								"location":"1:45000-96000:-1"}</a> <a
								href='api/genes/query?query={"genome":"homo_sapiens", "location":["1:45000-52000","1:60000-96000"]}'>api/genes/query?query={"genome":"homo_sapiens",
								"location":["1:45000-52000","1:60000-96000"]}</a>
						</code>
					</p>

					<h5>Controlling output</h5>

					<p>
						Adding different fields: <br />
						<code>
							<a
								href='api/genes/query?query={"name":"BRCA2","genome":"homo_sapiens"}&fields=genome,name,description,start,end'>api/genes/query?query={"name":"BRCA2","genome":"homo_sapiens"}&amp;fields=genome,name,description,start,end</a>
						</code>
					</p>

					<p>
						Returning results as a 2D array: <br />
						<code>
							<a
								href='api/genes/query?query={"name":"BRCA2","genome":"homo_sapiens"}&array=true&fields=genome,name,description,start,end'>api/genes/query?query={"name":"BRCA2","genome":"homo_sapiens"}&amp;array=true&amp;fields=genome,name,description,start,end</a>
						</code>
					</p>

					<p>
						Adding fields from a sub-object: <br />
						<code>
							<a
								href='api/genes/query?query={"name":"BRCA2","genome":"homo_sapiens"}&fields=genome,name,description,start,end,transcripts.name,transcripts.biotype'>api/genes/query?query={"name":"BRCA2","genome":"homo_sapiens"}&amp;fields=genome,name,description,start,end,transcripts.name,transcripts.biotype</a>
						</code>
					</p>

					<p>
						Adding all fields: <br />
						<code>
							<a
								href='api/genes/query?query={"name":"BRCA2","genome":"homo_sapiens"}&fields=*'>api/genes/query?query={"name":"BRCA2","genome":"homo_sapiens"}&amp;fields=*</a>
						</code>
					</p>

					<h5>Joining other data types</h5>

					<p>
						Joining to homologues: <br />
						<code>
							<a
								href='api/genes/query?query={"name":"BRCA2","genome":"homo_sapiens"}&fields=["name","description",{"homologues":["division","name"]}]'>api/genes/query?query={"name":"BRCA2","genome":"homo_sapiens"}&amp;fields=["name","description",{"homologues":["division","name"]}]</a>
						</code>
					</p>

					<p>
						Joining to homologues and restricting further to just primates: <br />
						<code>
							<a
								href='api/genes/query?query={"name":"BRCA2","genome":"homo_sapiens","homologues":{"lineage":"9443"}}&fields=["name","description",{"homologues":["division","name"]}]'>api/genes/query?query={"name":"BRCA2","genome":"homo_sapiens","homologues":{"lineage":"9443"}}&amp;fields=["name","description",{"homologues":["division","name"]}]</a>
						</code>
					</p>

					<p>
						Joining to variants: <br />
						<code>
							<a
								href='api/genes/query?query={"name":"BRCA2","genome":"homo_sapiens"}&fields=["name","description",{"variants":["id","chr","start","stop"]}]'>api/genes/query?query={"name":"BRCA2","genome":"homo_sapiens"}&amp;fields=["name","description",{"variants":["id","chr","start","stop"]}]</a>
						</code>
					</p>

					<p>
						Joining to variants and showing only genes with variants: <br />
						<code>
							<a
								href='api/genes/query?query={"name":"BRCA2","genome":"homo_sapiens","variants":{"inner":1}}&fields=["name","description",{"variants":["id","chr","start","stop"]}]'>api/genes/query?query={"name":"BRCA2","genome":"homo_sapiens","variants":{"inner":1}}&amp;fields=["name","description",{"variants":["id","chr","start","stop"]}]</a>
						</code>
					</p>

					<p>
						Joining to variants and showing counts only: <br />
						<code>
							<a
								href='api/genes/query?query={"name":"BRCA2","genome":"homo_sapiens"}&fields=["name","description",{"variants":["count"]}]'>api/genes/query?query={"name":"BRCA2","genome":"homo_sapiens"}&amp;fields=["name","description",{"variants":["count"]}]</a>
						</code>
					</p>

					<p>
						Joining to variants with given SO term: <br />
						<code>
							<a
								href='api/genes/query?query={"name":"BRCA2","genome":"homo_sapiens","variants":{"annot":{"ct":{"so":"1627"}}}}&fields=["name","description",{"variants":["id","chr","start","stop"]}]'>api/genes/query?query={"name":"BRCA2","genome":"homo_sapiens","variants":{"annot":{"ct":{"so":"1627"}}}}&amp;fields=["name","description",{"variants":["id","chr","start","stop"]}]</a>
						</code>
					</p>

					<p>
						Joining to expression data: <br />
						<code>
							<a
								href='api/genes/query?query={"name":"BRCA2","genome":"homo_sapiens"}&fields=["name","description",{"expression":["experimentType","expressionLevel"]}]'>api/genes/query?query={"name":"BRCA2","genome":"homo_sapiens"}&amp;fields=["name","description",{"expression":["experimentType","expressionLevel"]}]</a>
						</code>
					</p>

					<h5>Faceting</h5>

					<p>
						To facet by genome: <br />
						<code>
							<a href='api/genes/query?query={"name":"BRCA2"}&facets=genome'>api/genes/query?query={"name":"BRCA2"}&amp;facets=genome</a>
						</code>
					</p>

					<p>
						To facet by genome and biotype: <br />
						<code>
							<a href='api/genes/query?query={"name":"BRCA2"}&facets=genome'>api/genes/query?query={"name":"BRCA2"}&amp;facets=genome</a>
						</code>
					</p>

					<h5>Sorting</h5>

					<p>
						To sort by genome: <br />
						<code>
							<a
								href='api/genes/query?query={"GO_expanded":"GO:0006302"}&fields=genome,name,start,end&sort=genome'>api/genes/query?query={"GO_expanded":"GO:0006302"}&fields=genome,name,start,end&amp;sort=genome</a>
						</code>
					</p>

					<p>
						To facet by genome and start: <br />
						<code>
							<a
								href='api/genes/query?query={"GO_expanded":"GO:0006302"}&fields=genome,name,start,end&sort=genome,start'>api/genes/query?query={"GO_expanded":"GO:0006302"}&fields=genome,name,start,end&amp;sort=genome,start</a>
						</code>
					</p>

					<p>
						To facet by genome and start (descending): <br />
						<code>
							<a
								href='api/genes/query?query={"GO_expanded":"GO:0006302"}&fields=genome,name,start,end&sort=genome,-start'>api/genes/query?query={"GO_expanded":"GO:0006302"}&amp;fields=genome,name,start,end&amp;sort=genome,-start</a>
						</code>
					</p>

					<h5>Pagination</h5>

					<p>
						To show the first 10 entries: <br />
						<code>
							<a
								href='api/genes/query?query={"name":"BRCA2","lineage":"40674"}&limit=10'>api/genes/query?query={"name":"BRCA2","lineage":"40674"}&amp;limit=10</a>
						</code>
					</p>

					<p>
						To show the next 10 entries: <br />
						<code>
							<a
								href='api/genes/query?query={"name":"BRCA2","lineage":"40674"}&limit=10&offset=10'>api/genes/query?query={"name":"BRCA2","lineage":"40674"}&amp;limit=10&amp;offset=10</a>
						</code>
					</p>

					<p>
						NB: Performance is poor for very deep pagination - please use
						<code>/fetch</code>
						instead!
					</p>

					<h4>
						<code>/fetch</code>
					</h4>

					<p>
						Retrieving all results as JSON: <br />
						<code>
							<a href='api/genes/fetch?query={"name":"BRCA2"}'>api/genes/fetch?query={"name":"BRCA2"}</a>
						</code>
					</p>

					<p>
						Changing fields: <br />
						<code>
							<a
								href='api/genes/fetch?query={"name":"BRCA2"}&fields=genome,name,description,start,end'>api/genes/fetch?query={"name":"BRCA2"}&amp;fields=genome,name,description,start,end</a>
						</code>
					</p>

					<p>
						Returning results as a 2D array: <br />
						<code>
							<a
								href='api/genes/fetch?query={"name":"BRCA2","genome":"homo_sapiens"}&array=true&amp;fields=genome,name,description,start,end'>api/genes/fetch?query={"name":"BRCA2","genome":"homo_sapiens"}&array=true&amp;fields=genome,name,description,start,end</a>
						</code>
					</p>

					<p>
						Retrieving all fields: <br />
						<code>
							<a href='api/genes/query?query={"name":"BRCA2"}&fields=*'>api/genes/query?query={"name":"BRCA2"}&amp;fields=*</a>
						</code>
					</p>

					<p>
						Retrieving all results as XML (or pass the
						<code>Accept</code>
						properly as a request header): <br />
						<code>
							<a
								href='api/genes/fetch?query={"name":"BRCA2"}&accept=application/xml'>api/genes/fetch?query={"name":"BRCA2"}&amp;accept=application/xml</a>
						</code>
					</p>

					<p>
						Retrieving all results as CSV (or pass the
						<code>Accept</code>
						properly as a request header): <br />
						<code>
							<a href='api/genes/fetch?query={"name":"BRCA2"}&accept=text/csv'>api/genes/fetch?query={"name":"BRCA2"}&amp;accept=text/csv</a>
						</code>
					</p>

					<p>
						Retrieving genomic sequences for all results: <br />
						<code>
							<a
								href='api/genes/fetch?query={"name":"BRCA2"}&fields=[{"sequences":[]}]'>api/genes/fetch?query={"name":"BRCA2"}&amp;fields=[{"sequences":[]}]</a>
						</code>
					</p>

					<p>
						Retrieving genomic sequences for all results with additional
						100bp: <br />
						<code>
							<a
								href='api/genes/fetch?query={"name":"BRCA2","sequences":{"expand3_prime":"100","expand5_prime":"100}}&fields=[{"sequences":[]}]'>api/genes/fetch?query={"name":"BRCA2","sequences":{"expand3_prime":"100","expand5_prime":"100}}&amp;fields=[{"sequences":[]}]</a>
						</code>
					</p>

					<p>
						Retrieving CDS sequences for all results: <br />
						<code>
							<a
								href='api/genes/fetch?query={"name":"BRCA2","sequences":{"type":"cds"}}&fields=[{"sequences":[]}]'>api/genes/fetch?query={"name":"BRCA2","sequences":{"type":"cds"}}&amp;fields=[{"sequences":[]}]</a>
						</code>
					</p>

					<p>
						Retrieving protein sequences for all results: <br />
						<code>
							<a
								href='api/genes/fetch?query={"name":"BRCA2","sequences":{"type":"protein"}}&fields=[{"sequences":[]}]'>api/genes/fetch?query={"name":"BRCA2","sequences":{"type":"protein"}}&amp;fields=[{"sequences":[]}]</a>
						</code>
					</p>

				</div>
				<div class="tab-pane fade" id="aboutegs">
					<h3>About Ensembl Gene Search</h3>
					<p>
						A service for searching the annotation of genes from <a
							href="http://ensembl.org/">Ensembl</a> and <a
							href="http://ensemblgenomes.org/">Ensembl Genomes</a>
					</p>
				</div>
			</div>
		</div>

		<footer>

			<!-- Optional local footer (insert citation / project-specific copyright / etc here -->
			<!--
        <div id="local-footer" class="grid_24 clearfix">
      <p>How to reference this page: ...</p>
    </div>
        -->
			<!-- End optional local footer -->




			<div id="global-footer" class="grid_24">

				<nav id="global-nav-expanded">

					<div class="grid_4 alpha">
						<h3 class="embl-ebi">
							<a href="http://www.ebi.ac.uk/" title="EMBL-EBI">EMBL-EBI</a>
						</h3>
					</div>

					<div class="grid_4">
						<h3 class="services">
							<a href="http://www.ebi.ac.uk/services">Services</a>
						</h3>
					</div>

					<div class="grid_4">
						<h3 class="research">
							<a href="http://www.ebi.ac.uk/research">Research</a>
						</h3>
					</div>

					<div class="grid_4">
						<h3 class="training">
							<a href="http://www.ebi.ac.uk/training">Training</a>
						</h3>
					</div>

					<div class="grid_4">
						<h3 class="industry">
							<a href="http://www.ebi.ac.uk/industry">Industry</a>
						</h3>
					</div>

					<div class="grid_4 omega">
						<h3 class="about">
							<a href="http://www.ebi.ac.uk/about">About us</a>
						</h3>
					</div>

				</nav>

				<section id="ebi-footer-meta">
					<p class="address">EMBL-EBI, Wellcome Trust Genome Campus,
						Hinxton, Cambridgeshire, CB10 1SD, UK &nbsp; &nbsp; +44 (0)1223 49
						44 44</p>
					<p class="legal">
						Copyright &copy; EMBL-EBI 2016 | EBI is an Outstation of the <a
							href="http://www.embl.org">European Molecular Biology
							Laboratory</a> | <a href="/about/privacy">Privacy</a> | <a
							href="/about/cookies">Cookies</a> | <a href="/about/terms-of-use">Terms
							of use</a>
					</p>
				</section>

			</div>

		</footer>
	</div>
	<!--! end of #wrapper -->


	<!-- JavaScript at the bottom for fast page loading -->


	<!-- Grab Google CDN's jQuery, with a protocol relative URL; fall back to local if offline -->
	<script
		src="//ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
	<script src="js/bootstrap.min.js"></script>


	<!-- Your custom JavaScript file can go here... change names accordingly -->
	<script defer="defer"
		src="//www.ebi.ac.uk/web_guidelines/js/cookiebanner.js"></script>
	<script defer="defer" src="//www.ebi.ac.uk/web_guidelines/js/foot.js"></script>
	<!-- end scripts-->

	<!-- Google Analytics details... -->
	<!-- Change UA-XXXXX-X to be your site's ID -->
	<!--
<script>
  window._gaq = [['_setAccount','UAXXXXXXXX1'],['_trackPageview'],['_trackPageLoadTime']];
  Modernizr.load({
    load: ('https:' == location.protocol ? '//ssl' : '//www') + '.google-analytics.com/ga.js'
  });
</script>
-->

	<script
		src="https://ajax.googleapis.com/ajax/libs/jqueryui/1.11.4/jquery-ui.min.js"></script>
	<script type="text/javascript" charset="utf8"
		src="//cdn.datatables.net/1.10.11/js/jquery.dataTables.js"></script>
	<script type="text/javascript" charset="utf8"
		src="//cdnjs.cloudflare.com/ajax/libs/select2/4.0.2/js/select2.min.js"></script>
	<script src="js/search.js"></script>

	<!-- Prompt IE 6 users to install Chrome Frame. Remove this if you want to support IE 6.
chromium.org/developers/how-tos/chrome-frame-getting-started -->
	<!--[if lt IE 7 ]>
<script src="//ajax.googleapis.com/ajax/libs/chrome-frame/1.0.3/CFInstall.min.js"></script>
<script>window.attachEvent('onload',function(){CFInstall.check({mode:'overlay'})})</script>
<![endif]-->
</body>
</html>
